[2단계 - 로또(수동)] 쿼리치(정석윤) 미션 제출합니다.

meatsby opened the pull request:
안녕하세요 에단!

1단계 리뷰 너무 감사드립니다. 덕분에 컬렉션 반환 시 주의사항이나 네이밍 등 놓쳤던 디테일들에 대해 공부할 수 있었습니다!

1단계에서 남겨주셨던 피드백 적용 후, 기존에 있던 코드를 최대한 활용해 2단계를 구현해봤습니다!
추가적으로 개선할 점이나 잘못된 부분 지적해주시면 감사하겠습니다! 😊

bugoverdose committed:
docs(README): setup feature list

bugoverdose committed:
feat(Lotto): implement Lotto domain

- test size of numbers

bugoverdose committed:
feat(Lotto): numbers has sorted six random numbers

- test whether numbers are sorted ascending

bugoverdose committed:
feat(InputView): request user money input and parse to integer

bugoverdose committed:
feat(LottoGame): implement LottoGame domain

- init lottos with factory method
- test lottos size is equal to lottoCount argument

bugoverdose committed:
feat(study): how to join Integer List with stream

bugoverdose committed:
feat(OutputView): print lotto purchase info

bugoverdose committed:
feat(controller): run application with controller

bugoverdose committed:
fix(Lotto): create new array of numbers sublist

- implement toString for Lotto
- add study test on shuffle
- add test case on unique Lottos

bugoverdose committed:
feat(InputView): request winning numbers and bonus number

- refactor with readline

bugoverdose committed:
feat(validator): throw illegal arg exception on invalid number input

bugoverdose committed:
feat(validator): throw illegal arg exception on invalid input

- lotto nums should be in 1~45 range
- lotto nums should be unique

bugoverdose committed:
feat(validator): throw illegal arg exception on invalid input

- lotto nums should be in 1~45 range
- lotto nums should be unique

bugoverdose committed:
refactor(validator): exception messages to constants

- reduce depth with stream

bugoverdose committed:
test(LottoNumberValidators): add test cases for validators

- test pass and fail cases for all methods
- test validateLottoNumberRange with boundary values

bugoverdose committed:
feat(controller): validate range for winning numbers and bonus number input

bugoverdose committed:
feat(LottoNumber): value object for lotto number with range validation

bugoverdose committed:
feat(Lotto): create manual numbers input

bugoverdose committed:
feat(LottoResult): implement lotto result enum

- find matching enum with factory method and stream

bugoverdose committed:
feat(LottoReferee): implement lotto referee

- return LottoResult enum for numbers in lotto

bugoverdose committed:
feat(LottoNumber): replace int with LottoNumber objects, fix broken tests

bugoverdose committed:
feat(Lottos): implement lottos domain for list of lotto

bugoverdose committed:
feat(LottoGame): initialize LottoGame with Lottos and LottoReferee

bugoverdose committed:
feat(study): toMap can initialize Map with values of stream

bugoverdose committed:
refactor(Lottos): use factory method of to init with random lottos

bugoverdose committed:
feat(LottoGame): implement method for analyzing lottos and getting results statistics

- add test

bugoverdose committed:
feat(LottoResult): replace NONE with null, change enum order

- fix breaking tests

bugoverdose committed:
feat(LottoGame): Map to TreeMap for order keys ascending

bugoverdose committed:
feat(OutputView): print lotto result for each case

- add hasBonus property for LottoResult enum

bugoverdose committed:
feat(LottoGame): implement calculate lotto profit and price ratio

- move analyzing lottos logic to constructor

bugoverdose committed:
feat(OutputView): print profit and price ratio, round at third decimal place

bugoverdose committed:
test(LottoGame): calculateProfitRatio returns price by profit ratio

bugoverdose committed:
fix: fix conflict and merge

meatsby committed:
docs(README): rearrange feature list

meatsby committed:
refactor(OutputView): refactored printing method

- use printf for some print statements
- remove print method

meatsby committed:
feat(LottoNumber): implement cache structure for discrete LottoNumbers

- remove overridden methods
- implement method for getting LottoNumbers list
- add test case for additional method
- rename several variables and test methods

meatsby committed:
feat(UserBalanceValidator): implement validator for user balance

- add test for failing cases

meatsby committed:
style: rename variable and class name

meatsby committed:
refactor(LottoTickets): implement calculation for LottoTickets

meatsby committed:
style(LottoReferee): rename class name

meatsby committed:
refactor(WinningLotto): rearranged responsibility of validation

meatsby committed:
refactor(LottoNumber): rearranged responsibility of validation

meatsby committed:
refactor(LottoTicket): rearranged responsibility of validation

meatsby committed:
refactor(LottoNumber): rearranged responsibility of validation

meatsby committed:
refactor(LottoTickets): refactored parameter for constructor

- renamed test methods
- provision of display name

meatsby committed:
feat(UserBalance): implement user balance class for validation of initial user balance input

- add test cases

meatsby committed:
refactor(LottoResult): removed unnecessary variables and method

- provision of display name for test cases

meatsby committed:
refactor(LottoGame): renamed variables and methods

- remove magic number

meatsby committed:
refactor: restructured controller and view classes by reducing bonding with domain objects

meatsby committed:
refactor(LottoResult): removed unnecessary method

meatsby committed:
refactor(LottoTickets): refactored to return unmodifiable list

meatsby committed:
style: rename methods

meatsby committed:
refactor(UserBalance): conditional statement modification

- rename method and constants

meatsby committed:
style(LottoNumber): rename method and constant

meatsby committed:
refactor: refactored to return new collections

meatsby committed:
style: rename methods

meatsby committed:
docs(README): update feature list

meatsby committed:
feat: implement manual purchase

- refactor LottoTickets constructor and methods
- implement views for manual purchase
- reconstruct test methods

Laterality commented on src/main/java/controller/LottoController.java:
```diff
@@ -18,34 +18,60 @@
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("구매할 로또 수는 숫자여야 합니다.");
+        }
+    }
```
거의 동일한 코드가 `LottoController`, `LottoNumber`, `UesrBalance`에서 쓰이고 있습니다. 

과연 이 클래스들에서 할 필요가 있는 작업일까요? 이 작업을 수행하기 더 적절한 곳은 없을까요?

meatsby replied:
확실히 한 곳에서 처리해서 넘겨주는 것이 더 좋을 것 같네요.

해당 메서드를 활용해야 하는 객체가 상당해서 `controller` 나 `view` 에서 수행해주면 좋을 것 같은데 `view` 에서 데이터 가공을 해도 괜찮은가요?

현재까지 공부한 바로는 `view` 가 언제든 바뀔 수 있기 때문에 `domain` 에 맞춰서 `view` 를 제공해야 한다고 알고 있어요.
그래서 `view` 에서 `domain` 이 받기 편하게 끔 데이터를 가공해서 넘겨주도록 개선해봤는데 `view` 의 책임에 대한 지식이 부족하니 확신이 떨어지네요. 😥

혹시 추천해주실 만한 참고자료가 있다면 공유해주시면 감사하겠습니다!

Laterality replied:
결론 먼저 말씀드리면 각 영역의 책임을 어떻게 나누느냐에 따라 이 작업을 어느 영역에서 처리하는 것이 좋을지 결정할 수 있겠어요. 때문에 먼저 각 영역에서 어떤 책임을 가져야 할지 나눠보시면 좋겠습니다.

사용자가 뷰를 통해 값을 입력하면 그 값이 view, controller, model 영역을 통과하게 될 텐데요. 각 영역에서 자신의 관심사에 맞게 값을 처리한다고 보시면 돼요.

예를 들면, 모델 영역의 관심사는 전달받은 로또 번호들이 올바른 값인지, 숫자가 총 6개인지, 숫자 사이에 중복은 없는지 등등과 같은 부분일 테고요. 전달받은 로또 번호가 정수 타입인지 문자열 타입인지는 모델에서 별로 관심을 가질 만한 점은 아닐 테고, 모델 입장에서 값을 처리하기 가장 용이한 건 정수 타입이니 이 타입만 취급하면 됩니다.

그렇다면 말씀하신 대로 문자열 입력값에 대한 변환은 컨트롤러나 뷰에서 처리해야 할 테고요.

변환 작업은 컨트롤러나 뷰 어느 쪽에서 처리해도 문제 없다고 생각해요. 사용자의 입력값을 적절한 타입으로 반환하는 것을 뷰의 책임으로 볼 수도 있고, 뷰가 정제되지 않은 날 것(raw)의 입력값을 반환하면 적절하게 정제해서 모델 영역에 전달하는 것을 컨트롤러의 책임으로 보겠다면 이 또한 타당합니다.

Laterality commented on src/main/java/controller/LottoController.java:
```diff
@@ -18,34 +18,60 @@
+        if (lottoCount < 0 || lottoCount > userBalance.getUserBalance() / 1000) {
+            throw new IllegalArgumentException("수동 입력 가능 횟수는 0번 이상 최대 구입 가능 이하여야 합니다.");
+        }
+    }
```
1. 로또 가격(`1000`)은 매직 넘버네요.

2. 이 검증을 처리하기 더 적절한 곳은 없을까요?

meatsby replied:
매직 넘버 제거 완료했습니다!

아무래도 `LottoController` 에서 검증하는 건 어울리지 않는 것 같아요.
`UserBalance` 클래스에서 검증을 시도해보는 건 어떨까요?
아무래도 `manualLottoCount` 의 검증을 수행하려면 `userBalance` 값이 필요하기도 하고 추후 `autoLottoCount` 가 필요할 경우 `UserBalance` 클래스 내부에서 계산하는 것이 맞는 것 같아서요!

Laterality replied:
👍

Laterality commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -1,30 +1,30 @@
-        int lottoCount = userBalance.getUserBalance() / LOTTO_PRICE;
+    public void add(LottoTicket lottoTicket) {
+        lottoTickets.add(lottoTicket);
+    }
```
수동 로또를 직접 추가하기 위한 목적인 것으로 보이는데요. 컨트롤러를 살펴보기 전에는 사용 목적을 파악하기 어려워보입니다.

아래처럼 사용할 수 있다면 어떨까요?

```java
List<LottoTicket> manualLottos = new ArrayList();
// ... 수동 번호를 입력받아 manualLottos에 추가
lottoTickets.purchaseBy(userBalance, manualLottos);
```

meatsby replied:
훨씬 좋을 것 같아요!
대신 `purchaseBy()` 메서드로 수동과 자동을 모두 추가해주면 두 가지 일을 하는 것 같아서 `purchaseManualBy()` 와 `purchaseAutoBy()` 로 나눠봤는데 괜찮을까요?

Laterality replied:
💯

Laterality commented on src/main/java/domain/UserBalance.java:
```diff
@@ -11,21 +11,21 @@ public UserBalance(String value) {
 
     private void validateBalance(int balance) {
-        if (!(balance > 1000 && balance % 1000 == 0)) {
+        if (!(balance >= 1000 && balance % 1000 == 0)) {
```
매직 넘버가 여기도 있네요.

meatsby replied:
제거 완료!

Laterality reviewed (changes requested):
안녕하세요 쿼리치,

2단계 미션을 빠르게 완성해주셨네요.

몇가지 피드백을 남겨드렸어요. 참고 부탁드릴게요 :)

뷰, 컨트롤러, 모델의 책임 영역에 관해 고민해보시면 도움이 될 거에요.

meatsby committed:
refactor: rearranged responsibility of parsing number to view package

- reconstructed controller
- rearranged responsibility of lotto count validation to UserBalance class
- removed string lotto number constructor
- implement manual purchase method in LottoTickets class
- refactored test methods
- removed unnecessary print methods

Laterality commented on src/main/java/domain/UserBalance.java:
```diff
@@ -1,36 +1,47 @@
-    private static final String INVALID_CURRENCY_EXCEPTION_MESSAGE = "구입금액은 1000원 이상이어야 하며 1000원 미만일 수 없습니다.";
+    private static final String INVALID_BALANCE_EXCEPTION_MESSAGE = "구입금액은 1000원 이상이어야 하며 1000원 미만일 수 없습니다.";
+    private static final String INVALID_LOTTO_COUNT_EXCEPTION_MESSAGE = "수동 입력 가능 횟수는 0번 이상 최대 구입 가능 이하여야 합니다.";
+    public static final int LOTTO_PRICE = 1000;
```
`public`일 필요가 있을까요?

Laterality reviewed (approved):
안녕하세요 쿼리치,

피드백을 잘 반영해주셨네요.

추가로 피드백드릴 만한 내용은 없어서 이만 머지하도록 할게요. 

몇군데 코멘트 남겨드린 부분만 확인 부탁드립니다.

미션 고생 많으셨습니다 :)
